src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/MemoryContext.java

```diff
@@ -0,0 +1,257 @@
+        var path = packageAndClassName.replace('.', '/') + ".java";
+        var file = descriptor.sourceRootPath().resolve(path);
+
+        // Trivial case: no matching source file exists
+        if (Files.notExists(file)) return null;
+
+        // Compile source file (unit) with similar options as the program.
+        var opts = options.forSubsequentCompilations();
+        var unit = standardFileManager.getJavaFileObject(file);
+        var task = compiler.getTask(out, memoryFileManager, null, opts, null, List.of(unit));
+
+        var ok = task.call();
+        if (!ok) {
+            var fault = new Fault(Errors.CompilationFailed);
+            // Don't throw fault - fail fast!
+            out.println(fault.getMessage());
+            System.exit(2);
+        }
+
+        // The memory file manager stored bytes in the context map, indexed by the class names.
+        return inMemoryClasses.get(name);
+    }
+
+    /**
+     * Load main entry-point class.
+     *
+     * @param parent the class loader to be used as the parent loader
+     * @param mainClassName the fully-qualified name of the application class to load
+     * @return class object representing the desired class
+     * @throws ClassNotFoundException if the class cannot be located
+     * @throws Fault if a modular application class is in the unnamed package
+     */
+    Class<?> loadApplicationClass(ClassLoader parent, String mainClassName) throws ClassNotFoundException, Fault {
+        var moduleInfoBytes = inMemoryClasses.get("module-info");
+        if (moduleInfoBytes == null) {
+            // Trivial case: no compiled module descriptor available, no extra module layer required
+            var memoryClassLoader = new MemoryClassLoader(inMemoryClasses, parent, null, descriptor, this::compileJavaFileByName);
+            return Class.forName(mainClassName, true, memoryClassLoader);
+        }
+
+        // Ensure main class resides in a named package.
+        var lastDotInMainClassName = mainClassName.lastIndexOf('.');
+        if (lastDotInMainClassName == -1) {
+            throw new Fault(Errors.UnnamedPkgNotAllowedNamedModules);
+        }
+
+        var bootLayer = ModuleLayer.boot();
+        var parentLayer = bootLayer;
+        var parentLoader = parent;
+
+        // Optionally create module layer for all modules on the module path.
+        var modulePathFinder = createModuleFinderFromModulePath();
+        var modulePathModules = modulePathFinder.findAll().stream().map(ModuleReference::descriptor).map(ModuleDescriptor::name).toList();
+        if (!modulePathModules.isEmpty()) {
+            var modulePathConfiguration = bootLayer.configuration().resolveAndBind(modulePathFinder, ModuleFinder.of(), Set.copyOf(modulePathModules));
+            var modulePathLayer = ModuleLayer.defineModulesWithOneLoader(modulePathConfiguration, List.of(bootLayer), parent).layer();
```

Comment:
Since you don't need the Controller then it may be simpler to use just `bootLayer.defineModulesWithOneLoader(modulePathConfiguration, parent);`

Reply:
Will change the line as suggested.